ENH: Refactor observations in MRML nodes (#8350)
Slicer/Slicer: Docs/developer_guide/advanced_topics.md

## Event broker

The basic idea of the EventBroker is to avoid the following kind of code that shows up in GUIs and other classes that depend on events generated by other classes:

```cpp
node->AddObserver(vtkCommand::ModifiedEvent, callbackCommand)
```

### Problem

The problems with this:
- node 'owns' the observer, but the callbackCommand is opaque so it doesn't know anything about what will happen when the event is invoked
- the GUI needs to explicitly remove the observer before it is destroyed
- node is not introspectable; you cannot get a list of observers on the node (all data and methods are private)
- there's no easy way to know what side effects will happen for any Set call (either a priori or experimentally).
- there's no way to collapse events or disable them

### Solution

The EventBroker solve this by introducing a singleton that manages all observations.

The EventBroker is a singleton, available as:

```cpp
vtkEventBroker *broker = vtkEventBroker::GetInstance();
```

with the broker, you can make a call like the following:

```cpp
broker->AddObservation(node, vtkCommand::ModifiedEvent, this, callbackCommand);
```

where `node` is the subject of the observation.

> The broker is also available for use from Python, but for adding observers in Python, it is recommended to use the higher-level `VTKObservationMixin` instead.

Author: "Tip" is OK, as it is correct to use the broker from Python

The broker does the following:

> - adds DeleteEvent observers to both `node` and `this` so it can remove the observation automatically when either side is destroyed

Reviewer: ```suggestion
- adds `DeleteEvent` observers to both `node` and `this` so it can remove the observation automatically when either side is destroyed
```

- keeps an introspectable list of all observers it knows about
- has an option to keep a log of all event invocations for debugging and performance analysis
- has an option to turn off all event invocations
- has an _asynchronous_ option to queue all event invocations and invoke them later (off by default)
- has option to collapse redundant events in the queue

Future options include:
- add a timer to log the amount of time taken to process each event
- specify that some observations must be handled synchronously (e.g. for progress events that shouldn't be collapsed)

> See the file [vtkEventBroker API documentation](https://apidocs.slicer.org/main/classvtkEventBroker.html) for more information on logging options and introspection features.

Reviewer: ```suggestion
See the file [vtkEventBroker API documentation](slicerapidocs:classvtkEventBroker.html) for more information on logging options and introspection features.
```

For reference:
* https://github.com/Slicer/Slicer/pull/8176

### How to work with the vtkObserverManager

Discussion of Jan 27 2012:

> [...] the problem wasn't that the Event Broker was used instead of going through vtkObserverManager, but that the vtkObject::AddObserver() was used with the callback command of the observer manager (that was later deleted which made the client data of the callback dirty when the event was fired).

Reviewer: ```suggestion
[...] the problem wasn't that the Event Broker was used instead of going through `vtkObserverManager`, but that the `vtkObject::AddObserver()` was used with the callback command of the observer manager (that was later deleted which made the client data of the callback dirty when the event was fired).
```

We should eventually review the entire "observation" mechanism, there are currently 3 ways of observing vtkObjects:

> - via the vtkObserverManager API (please everybody use that one for now)
> - via the EventBroker directly
> - via vtkObject::AddObserver directly

Reviewer: ```suggestion
- via the `vtkObserverManager` API (please everybody use that one for now)
- via the `EventBroker` directly
- via `vtkObject::AddObserver` directly
```

The basic idea of the EventBroker is to avoid the following kind of code that shows up in GUIs and other classes that depend on events generated by other classes:

```cpp
node->AddObserver(vtkCommand::ModifiedEvent, callbackCommand)
```

### Problem

The problems with this:
- node 'owns' the observer, but the callbackCommand is opaque so it doesn't know anything about what will happen when the event is invoked
- the GUI needs to explicitly remove the observer before it is destroyed
- node is not introspectable; you cannot get a list of observers on the node (all data and methods are private)
- there's no easy way to know what side effects will happen for any Set call (either a priori or experimentally).
- there's no way to collapse events or disable them

### Solution

The EventBroker solve this by introducing a singleton that manages all observations.

The EventBroker is a singleton, available as:

```cpp
vtkEventBroker *broker = vtkEventBroker::GetInstance();
```

with the broker, you can make a call like the following:

```cpp
broker->AddObservation(node, vtkCommand::ModifiedEvent, this, callbackCommand);
```

where `node` is the subject of the observation.

:::{tip}

> The broker is also available for use from Python, but for adding observers in Python, it is recommended to use the higher-level :class:`~slicer.util.VTKObservationMixin` instead.

Author: ```suggestion
The broker is also available for use from Python, but for adding observers in Python, it is recommended to use the higher-level `slicer.util.VTKObservationMixin` instead.
```
